lib/vector-core/buffers/src/disk_v2/ledger.rs

```diff
@@ -0,0 +1,334 @@
+use std::{
+    fmt, io,
+    path::{Path, PathBuf},
+    sync::atomic::{AtomicU16, AtomicU64, Ordering},
+    time::{Duration, Instant},
+};
+
+use bytecheck::CheckBytes;
+use bytes::BytesMut;
+use crossbeam_utils::atomic::AtomicCell;
+use fslock::LockFile;
+use memmap2::{MmapMut, MmapOptions};
+use rkyv::{with::Atomic, Archive, Serialize};
+use snafu::{ResultExt, Snafu};
+use tokio::{fs::OpenOptions, io::AsyncWriteExt, sync::Notify};
+
+use super::{backed_archive::BackedArchive, ser::SerializeError};
+
+#[derive(Debug, Snafu)]
+pub enum LedgerLoadCreateError {
+    #[snafu(display("ledger I/O error: {}", source))]
+    Io { source: io::Error },
+    #[snafu(display(
+        "failed to lock buffer.lock; is another Vector process running and using this buffer?"
+    ))]
+    LedgerLockAlreadyHeld,
+    #[snafu(display("failed to deserialize ledger from buffer: {}", reason))]
+    FailedToDeserialize { reason: String },
+    #[snafu(display("failed to serialize ledger to buffer: {}", reason))]
+    FailedToSerialize { reason: String },
+}
+
+/// Ledger state.
+///
+/// Stores the relevant information related to both the reader and writer.  Gets serailized and
+/// stored on disk, and is managed via a memory-mapped file.
+///
+/// # Warning
+///
+/// - Do not add fields to this struct.
+/// - Do not remove fields from this struct.
+/// - Do not change the type of fields in this struct.
+/// - Do not change the order of fields this struct.
+///
```

Comment:
[This place is not a place of honor... no highly esteemed deed is commemorated here... nothing valued is here.](https://en.wikipedia.org/wiki/Long-time_nuclear_waste_warning_messages)